src/util/messages.ml

```diff
@@ -3,6 +3,137 @@ open Pretty
 open GobConfig
 module GU = Goblintutil
 
+module ArrayOOB =
+struct
+  type t =
+    | PastEnd
+    | BeforeStart
+    | Unknown
+
+  let show e =
+    match e with
+    | PastEnd -> "Index is past the end of the array."
+    | BeforeStart -> "Index is before start of the array."
+    | Unknown -> "Not enough information about index."
+end
+
+module UndefinedBehavior =
+struct
+  type t =
+    | ArrayOutOfBounds of ArrayOOB.t
+    | NullPointerDereference
+    | UseAfterFree
+
+  let show e =
+    match e with
+    | ArrayOutOfBounds e -> Printf.sprintf "[Array out of bounds] %s" (ArrayOOB.show e)
+    | NullPointerDereference -> "[Null pointer dereference]"
+    | UseAfterFree -> "[Use After Free]"
+end
+
+module BehaviorEvent =
+struct
+  type t =
+    | Undefined of UndefinedBehavior.t
+    | Implementation
+    | Machine
+
+  let show e =
+    match e with
+    | Undefined u -> Printf.sprintf "[Undefined] %s" (UndefinedBehavior.show u)
+    | Implementation -> "[Implementation]"
+    | Machine -> "[Machine]"
+end
+
+module IntegerEvent =
+struct
+  type t = Overflow | DivByZero
+end
+
+module CastEvent =
+struct
+  type t = TypeMismatch
+end
+
+module ArrayEvent =
+struct
+  type t = OutOfBounds of int*int
+end
```

Comment:
It's quite clear that these aren't properly used if they don't have `show` implemented and everything is still fine.

Reply:
yes, these are just prepared here to be extended as we weren't exactly sure where to use them or what information would be relevant to pass into them. Plus for example, we now have _Array out of bounds_ under _Undefined behavior_ and I actually don't know which one should it be - whether array out of bounds should be in this Array event or whether it should belong under the undefined behavior which it also is.

Reply:
I would put it in undefined behavior. Most of our warnings prob. are about UB. Which ones aren't besides asserts?
If we had some warnings in `Array` and `UB.Array`, it'd be cumbersome to filter for both.
If we have such cases, we could let `UB.Array` imply `Array` for filtering.